### ISSUE TITLE
Rename FloatValue to FloatSeq?

### ISSUE THREAD

**Issue by ngeiswei** [opencog/atomspace#1880]
ProtoAtoms such as `FloatValue`, `LinkValue`, etc actually are sequences. For instance `FloatValue` is
```
std::vector<double>
```
Why not rename them `FloatSeq`, etc?

It would be consistent with their meanings, and leave some rooms for other names the day someone wants to introduce single values as opposed to sequences.

Also, maybe `LinkValue` should be renamed `AtomSeq`.

**Comment by vsbogd** [comment]
As far as I understand it is named so because it is a parent class of the TruthValue. And TruthValue in turn is 2 floats actually. I agree that name is confusing, but I think we should better change implementation instead of renaming. 

Thus FloatValue should contain single float. TruthValue should contain two float fields instead of vector. And they should not inherit each other. And so on. I am not sure what is impact on compatibility though.

**Comment by ngeiswei** [comment]
@vsbogd SimpleTruthValue has 2 values but other truth value types such as CountTruthValue have different numbers of values, this is thus why TruthValue inherits from `FloatValue` (note saying it is the best design, I simply don't know, but just to explain why it inherits from `FloatValue` as opposed to say `FloatPair`).

**Comment by vsbogd** [comment]
@ngeiswei, yes, sure, there are other values in this hierarchy which inherit FloatValue and for some of them using vector as internal representation makes sense. 

What I am trying to say in other words:
- my opinion TruthValue is not FloatSequence, so simple renaming will introduce another sort of confusion
- we still need FloatValue to represent single float instead of vector

**Comment by ngeiswei** [comment]
@vsbogd agreed, for instance generalized distributional truth values are not FloatSeq (more like float counter, or something).

Also agreed about having FloatValue eventually representing single floating number rather than vector, just need to define it strictly after this first round of renaming to not create lots of confusion.

**Comment by rTreutlein** [comment]
Is anybody working on this? If not I might do it as I could use a FloatValue that's just a single float.

**Comment by ngeiswei** [comment]
Let's wait for @linas' feedback.

And let me stress again that turning FloatValue into a single value will have to be in a separate PR, doing otherwise would be way to0O DANGEROUS!

Also beware that it might affect the opencog repo...

**Comment by linas** [comment]
Hi, Sorry for late reply. 

1) Regarding the renaming: Now that you know its actually a sequence, what's the problem? Are you afraid that you'll wake up tomorrow morning, having forgotten that its a sequence?

2) Regarding "single" values: what's wrong with a sequence of length one? Does something get easier or harder, more efficient or less efficient, by introducing "single" values?

3) Explanation of why its a sequence: the primary, number-one driver of the value design is compute-efficiency: minimizing cpu cycles, minimizing ram usage, and for storage to disk (i.e. to sql), minimizing size.  Second is implementation complexity (although, it is usually the case that small, fast things are also not complex).  A distant third is human-usability. Again: the goal here is to make it easy for algorithms (not humans) to access and manipulate data.  Since algorithms are stupid (algorithms don't have general intelligence), this is best accomplished by keeping data structures and API's as simple as possible.  I'm trying to avoid a system with lots of frills and features. 

3a) Efficiency: given the above, if you need to store more than one float (for truth values, attention values) then how can you do this efficiently?  Try to write down every possible way of doing this, and then try to count CPU cycles and RAM usage. When I did this, the answer that came back was "use c++ std::vector". If there is a better way, then it's outside of what I can currently imagine.

4) Naming of  `LinkValue`: Yeah. That was hard. The most complicated part of 3) above was here: should values be JSON like? Should there be a difference between sets and arrays? Should there be linked lists?  Should there be IntValues and CharValues? The JS in JSON stands for JavaScript, and the atomspace really is a whole lot like Javascript. A huge amount like javascript.  If you don't know JS, you should take a month and code up some non-trivial app in it. You'll see what I mean.

4a) Clearly, there was a need for arrays-of-values. Or arrays of atoms. Or something. What's the difference between an array of atoms and an array of values?  Ugh. So what possible candidate names are there?  I listed all of those, and they were all ugly.  So `LinkValue` was the least odious of them.

**Comment by linas** [comment]
Anyway: to continue with point 2) -- re single-value float-values: again what's the point? what's the problem? why should you even care? If you want to store a single float, you can. Nothing is stopping you.  Perhaps you think that it's more efficient to not use `std::vector<double>`? Barely, hardly.  Try it, measure it. A microscopic amount of the performance goes there; most of it is lost in the python/guile bindings, most of the rest in the atomsapce, and then a decent chunk to look up value-by-key.  If you get rid of `std::vector<double>` then you need to write a bunch of special-case code in the SQL backend, to handle each and every different thing you replace it with. Yuck. Likewise for whatever RESTful interface, or whatever zermq interface, or whatever you come up with. The more data formats there are, the more code you have to write, the more bugs you introduce, the more unit-tests to catch them... yuck. 

Under the covers, the vector can store a single float, and to you, the user, why should it matter what happens under the covers? why should you care?  How much time do you spend examining the implementation details of GCC?  Worrying about how they implemented things? As long as GCC is "good enough", and "does the right thing", you mostly don't care about the assembly code falling out the back of it.  Same here: you can represent a single float just fine, why should you care about how it happened, under the covers?

**Comment by ngeiswei** [comment]
@linas, I'm personally neutral on single value vs singleton vector, just consider that a single value takes at most 8 bytes, while an empty vector already takes 16 bytes. I would agree though that we don't need to worry about that before it becomes an actual problem.

However
1. We need to allow room for single value to happen if we ever need them (maybe once the pattern matcher supports boolean predicates we'll want that, not sure, just an example).
2. The names `FloatValue` and `LinkValue` are misleading, I think even for math literates, they may resolve the apparent inconsistency quicker but still. It's not just me saying that, I'm regularly asked about them...

I'm obviously open to other names than `FloatSeq`, etc. like `FloatValues` or such.

Regarding renaming `LinkValue` (I still don't understand why `Link` was used), it's actually a vector of proto atoms, not regular Atom, so `AtomSeq` or such would be misleading too, maybe it could be `ProtoAtomSeq` or `ProtoAtomValues`.

**Comment by ngeiswei** [comment]
Obviously one may ask, if `TruthValue` then why not `FloatValue`?

My answer is: the notion of truth is abstract, it can be a boolean to an infinite order probability, nobody really knows (even Sam Harris and Jordan Peterson disagree on its definition). The notion of float however is rather concrete...

**Comment by ngeiswei** [comment]
But yeah these objects are call generally called values, as opposed to atoms (though they can be as well), so maybe we can rename `FloatValue` to `FloatSeqValue` or `FloatsValue`.

**Comment by andre-senna** [comment]
Linas: "I'm trying to avoid a system with lots of frills and features."

I think this is reason enough to not having a special case for single value implementation. Yes, there's a waste of RAM when you need to keep a single float in the vector but trying to optimize something without knowing it's an actual performance problem is an anti-pattern.

Linas: "Now that you know its actually a sequence, what's the problem? Are you afraid that you'll wake up tomorrow morning, having forgotten that its a sequence?"

IMO If the name confused an experimented developer it'll probably confuse newcomers as well. I see no reason to keep bad naming here.

**Comment by noskill** [comment]
@ngeiswei Another name suggestion -  FloatArray or FloatVec

**Comment by linas** [comment]
OK, `FloatArray` or `FloatArrayValue` is an OK name, I guess. The second name reminds you that it's not an atom. The distinction between atoms and values is very important -- if users have any confusion about that, things go down-hill very quickly.

The problem with the name "vector" is that comp-sci/programming vectors are not actually 'vectors' in the mathematical sense; they fail to obey the axioms for a vector. 

@vsbogd there are no users currently, or foreseen, for single floats. Truth values are all 2,3,or 4 floats. Attention values are 4 floats. All of the various ad-hoc values in the language learning code are 3 or 4 or 5 floats, for example, count, normalized-count (probability), log-probability and p log p. Its useful to store all of these at once, it avoids repeated calculations (e.g. log p might be needed thousands of times).

The space-savings of not using an `std::vector` is miniscule: as @ngeiswei  pointed out, its 16 bytes, which is about 1% of the total atom size, and maybe half of that if one also counts the values.  The performance overhead is also trite: one can do 100 million `std::vector` accesses a second; one can do 20K python/scheme accesses a second - the performance impact is 1/100th of one percent.

Worse, if you decide to store two distinct floats, instead of an array of length two, the overhead is huge: storing one value takes something like 150 or 200 bytes, because of the key-value system uses an `std::map<smart-ptr, smart-ptr>` and each smart-pointer is maybe 40 or 50 bytes, and the std-map entry itself (a rb-tree) is another 40 or 50 bytes, and the value itself is 20 or 40 or more bytes.  So storing two distinct floats instead of an array uses about 10x or 20x more storage.  There's just no way to win storing singe floats.

This analysis is what I meant when I said "think of all the other ways you could implement this". All the other ways I could think of were losers; this way seemed like the best way.  If there's a better way, that would be great, but I don't know what that better way is.

**Comment by linas** [comment]
... or `FloatSeqValue` is OK, I guess.  I'm just not very concerned about the newbie not realizing that this is an array, instead of a single value.  By contrast, I am very very very concerned that the newbie has a very clear idea of the difference between atoms and values, and the fact that they have very different performance profiles, and very different mutability and associativity properties.  By the time they figure this out, the fact that a `FloatValue` is actually an array will be kind-of a trivial fact.  This is not the hard part ...

**Comment by ngeiswei** [comment]
I would vote for `FloatSeqValue` as well, because `Seq` tends to be used in the code to denote a list of a vector. And similarly I would go with `ProtoAtomSeqValue`.

@rTreutlein, I'm happy to have you do the change (`FloatValue`->`FlaotSeqValue`, etc), however I do agree that, as of today, introducing a single value doesn't seem necessary. Could you describe the reason you want to introduce single values?

**Comment by rTreutlein** [comment]
@ngeiswei I am using single values for the distributional values but I can just use a [0] when accessing the value. It just seems less clean then if I knew that there could only be 1 number in the FloatValue.

I can still work on the renaming but if it is just that then i don't mind if somebody else wants to do it.

**Comment by ngeiswei** [comment]
Maybe for now work with a singleton vector, the representation for a DV might change anyway and we can revisit this sort of things when it's appears that introducing a single value provides a practical benefit.

**Comment by ngeiswei** [comment]
> Adding DistributionalValue type can make finding all usages of DV value simpler in future.

I think we'll eventually want to do that, yeah. AtomSpace-MOSES also needs some form of distributional value.

**Comment by rTreutlein** [comment]
There will be a DistributionalValue Type and the way it is right now also a ConditionalDistributionalValue Type.

**Comment by linas** [comment]
If we are renaming things, I would like to get rid of the term "proto-atom". It has outlived it's purpose and conceptual usefulness. 

Ideally, I want the new name to begin with the letter "a".   can't think of one ... iota, dot, jot, speck ... infinitesimal ...

**Comment by linas** [comment]
@rTreutlein questions:

1) for dereferencing the [0], can't you just write a wrapper routine to just do that for you? For you, it's the API that matters, and not how it is implemented under-the-covers.  Please distinguish these concerns (api vs. implementation) as being distinct concerns.

2) A "distributional value" in the standard sense of probability theory would be a  bin-count (aka histogram, frequency count, etc.) which require N numbers - very naturally a sequence.  https://en.wikipedia.org/wiki/Histogram

One side-effect of making FloatValue a sequence was I figured it would make @ngeiswei very happy by giving him a very natural distributional truth value. i.e. a histogram.  So I'm confused by how a distribution can be just a single value??

**Comment by linas** [comment]
Just to be clear:
```
   FloatSeqValue  4 5 6
```
is much much more efficient than
```
LinkValue
         FloatValue 4
         FloatValue 5
         FloatValue 6
```
It's almost criminal to use LinkValue like this: it would bloat the atomspace, it would run 4x slower, kill database performance, etc.

**Comment by ngeiswei** [comment]
Regarding distributional value, a histogram is good but you do need to map the random variable co-domain partition to the bins. I would like the distributional value representation to be expressive enough that it can represent distributions not just over probabilities, and not just over equidistant bins, so I was thinking more like a map from anything to float.

**Comment by ngeiswei** [comment]
Instead of map it could also be 2 vectors, one `FloatSeqValue` and one `ValueSeqValue` linked by their indices. And one could establish a convention such that if the partition vector is missing, then it represents equidistant bins over [0, 1]. This remains to be decided I guess.

**Comment by ngeiswei** [comment]
Oh, what about `LinkValue` -> `SeqValue`?

**Comment by linas** [comment]
> regarding renaming proto-atom, what about Value?

Yes. Rename `ProtoAtom.h` to `Value.h`; rename directory `opencog/atoms/proto` to `opencog/atoms/value` -- although I'm tempted to rename directory to `avalue` so that the first three are abc: `avalue`, `base` and `core` which is the first steps of the actual hierarchy.

**Comment by linas** [comment]
> `LinkValue` -> `SeqValue`?

Well, *all* values are already sequences. So changing the name to say its a sequence is kind-of goofy.  And it's hard to envision a value that would not be a sequence -- they will always be sequences, this seems like an inescapable fact.

**Comment by linas** [comment]
Re: distributional values, see the comments I made today, on #833 and  #1884 -- For distributional TV's, I think you really want to have an OO API, rather than  a static data representation.  The reason is that static representations are hard to define, and rarely/never are general enough for all users. By contrast, the whole idea of OO design is that users can implement the objects however they feel like, as long as they honor the API contract, and the system still "just works".

Also, in case it is not clear, distributional TV's are just matrices, with rows and columns ...

**Comment by ngeiswei** [comment]
I finally understand why `LinkValue` is called this way. It is because it is like a Link (a sequence of outgoing atoms), except that it is a sequence of outgoing values, and it is itself a value.

I still think we need to rename `FloatValue` but now I think it could be `FloatLinkValue`.

So I would suggest 2 options

## Using `Link`
1. `FloatValue` -> `FloatLinkValue`
2. `StringValue` -> `StringLinkValue`

## Using `Seq`
1. `FloatValue` -> `FloatSeqValue`
2. `StringValue` -> `StringSeqValue`
3. `LinkValue` -> `SeqValue`

The reason I suggest such renaming is because I strong anticipate we'll want to introduce values are are not sequences (such as `BoolValue` once we introduce boolean predicates for the pattern matcher). And in fact there's already one such value, called `Valuation`. The fact that it has lost its `Value` suffix doesn't make less of a value since it inherits from `Value`.

**Comment by ngeiswei** [comment]
The problem with using `Link` is that it relates to the notion of atom type link, and although it is related to that notion in some sense I fear that
1. it's more confusing that enlighting
2. then should non-sequence value be called `NodeValue`, like `BoolNodeValue`

It seems to me simpler to depart from the notion of nodes and links for values and just use `Seq` for sequences. As value do not have the notion of incoming links, I think it is be better suited.

**Comment by ngeiswei** [comment]
I have created a PR for the `Seq` option (+ subsequent PRs for other repos).

I'm open to other renaming schemes, such as `FloatValue`->`FloatLinkValue`, but I feel strong on renaming `FloatValue` and such because again it is confusing and assuming that all values are sequences is incorrect. `TruthValue` are not a sequence of `Truth`, nor `DistributionalValue` a sequence of `Distributional`, nor `PtrValue` sequence of pointers (regardless of whether it makes it into the repo).

And again I see as likely the possibility of introducing some `BoolValue` for boolean predicates, and these will be used all over the place, so a byte (or int) vs a vector will have an impact (not mentioning the naming clarity advantage).

**Comment by linas** [comment]
I really basically don't understand the urge to rename this. It seems to be a "color of the bicycle shed" problem.

A much, much more important problem is figuring out how to transition `cog-evaluate` and all the related infrastructure so that it returns generic values, instead of truth values. This will have a lot of icky, sticky issues that will require resolution, and given how things are going, it's becoming quite very urgent.
 
I kind-of want to resist BooleanValue for very obvious reasons. Not only is it another bike-shed problem, but it encourages bad design and sloppy thinking.

**Comment by linas** [comment]
Let me describe the past, and the future.

In the beginning, there were five different kinds of TruthValue and AttentionValue, all of which are multi-component sequences (one of the variants has six floats in it.) So all values were *already* sequences, and so Float and String were not going to be any different: they would also be sequences, just like the existing ones.

The important thing was to get everyone to understand what a Value actually was, and that was not easy. There was a fair amount of resistance and push-back to the idea; it was not loved.  So I wanted all Values to be of the same kind: they would *all* be sequences; FloatValue was the obvious base class for TruthValue and AttentionValue, while NodeValue and LinkValue were a just sneaky way to increase the variety of things you could do.

I think it still makes sense to have them all be sequences, always. The reason for this is that the incremental overhead of using `std::vector` is tiny compared to the overhead of using smart-pointers, keys and `std:map`, all of which are large, bulky and slow.    Representing the two floats in a SimpleTruthValue in any other way is just crazy. Cache locality matters, even if you don't actually know what cache locality is, or how it affects the run-time of your programs.

I'm very concerned that by renaming things, newcomers to the project will get the "bright idea" to create values that are NOT based on `std::vector`.  If you rename `FloatValue` to `FloatSeq`, someone will decide that they should invent something "new and useful", called "FloatValue" which stores a single float. And that would be a monumentally stupid idea, and then all of a sudden it requires extremely long discussions about cache locality and smart pointers and `std::map`, and the fact that everything else is already a sequence, always, no matter what.   And that inventing new types requires altering existing infrastructure e.g. the SQL backend.  

We are already having this kind of an argument for the NN work, and its terrible. For whatever reason, our neural-net friends misunderstood what a Value is, and how to use it, and got tied up into a huge a knot over it. Even Ben feels obligated to weigh in, and its clear he didn't really understand what a Value is either.  Explaining it is hard.  It takes intellectual effort. Things got built in a certain way for a certain reason, and it appears to be really really hard to get people to understand WHY things were designed a certain way. 

Thus, one way to avoid crazy future arguments is to use a naming convention that prevents those future arguments from ever happening. All values are always sequences, always no matter what, from now until forever.   You don't have to put "seq" into the name, because it simply is not possible to have any other kind of value. For example: there is no TruthSeqValue!  All Truth values are already sequences. The last thing I want is the invention of the sequence of truth value sequences.   If you really need that, use LinkValue.

The names of things affect how people think about them. Give them the wrong name, people think about them incorrectly.  I am very worried that `FloatSeq` will cause users to think that there might be some values that are non-sequences. (there aren't any) or try to invent non-sequences (they shouldn't).

**Comment by linas** [comment]
Part of the confusion might be due to the fact that Values need not be static: they can be dynamically changing, and the way that they change is governed by things outside of the atomspace. 

A good example if this is here: https://github.com/opencog/atomspace/blob/master/examples/atomspace/stream.scm it demonstrates vectors that are time-varying; one of the vectors is 24-dimensional. It is passed, as a vector, through PlusLink and TimesLink, which can perform basic math on it.  This example was created as a prototype for the neural-net interfaces to the atomspace; the idea was to replace `RandomStream` by the actual neural net.

**Comment by ngeiswei** [comment]
I disagree, @linas. Although truth value is implemented as a sequence, it is in spirit a single entity. And I think it's just a matter of time before singleton values are gonna be used as single entity as well, which will lead us, out of mere performance gain, to define non-sequential values.

But since we're not there yet. I guess let's just have time do the talk.

**Comment by linas** [comment]
Non-sequential values are a performance loss. Again, think about cache locality for a little bit. https://en.wikipedia.org/wiki/Locality_of_reference

**Comment by vsbogd** [comment]
@linas, but keeping `vector` inside `FloatValue` will not prevent someone from using `FloatValue` to keep single `float` number in this vector. So even if it is more optimal to keep big coupled set of floats in `FloatValue` nothing prevents people to keep it in `ListValue` of `FloatValue`, this decision is still on implementer side. 

And I think it should be on implementer side. If one needs sets of floats he can use `vectors` to keep values, if one need a single value why not allow him keeping it in separate `xxxValue` class?
